persist: slightly less WIP beginnings of example of how kafka upsert might work

aljoscha opened the pull request:
There are still some races in there and we duplicate the records of the last timestamp binding. But: you can restart this and the source will pick up where it left off, roughly.

I disabled concatenation of the restored upsert stream for now because the example would produce duplicates before because both the `persisted_source()` and the `persist()` emitted all the updates.

aljoscha committed:
persist: more WIP on "kafka" upsert example

There are still some races in there and we duplicate the records of the
last timestamp binding.

danhhz commented on src/persist/examples/kafka_upsert.rs:
```diff
@@ -49,8 +51,11 @@ fn construct_persistent_kafka_upsert_source<G: Scope<Timestamp = u64>>(
-    for ((k, v), _, _) in out_read.snapshot()?.read_to_end_flattened()? {
-        prev_value_by_key.insert(k, v);
+    for ((k, v), ts, diff) in out_read.snapshot()?.read_to_end_flattened()? {
+        if ts > start_ts {
```
shouldn't this be `<` (or maybe `<=`)? to read the data "as of" a timestamp, you need the data below it, no?

aljoscha replied:
The construction is slightly unintuitive, but we're filtering out anything that is strictly above `start_ts` by `continue`'ing, so we should be taking into account all the data that is `<= start_ts`.

danhhz replied:
🤦 carry on

danhhz reviewed (approved):
awesome! i didn't look at the operator impls closely, but they seemed good at a glance

one thing this immediately makes me notice that we're paying the persist latency twice before we finally emit a record (i think?). i wonder if there's some way to drop this down to once

aljoscha committed:
persist: add kafka_upsert_v2.rs as copy of kafka_upsert.rs

aljoscha committed:
persist: WIP on monolithic persistence source

aljoscha committed:
kafka: allow manual override of `enable.idempotence`

We hard-enable this in our producer but some users want to produce to
systems that don't support Producer IDs.

Allowing manual override will unblock them, but this setting is
potentially dangerous because it can (will?) lead to duplicate records
being produced.

aljoscha committed:
WIP
